### PR TITLE
Coordinate frames that include solar differential rotation

### LINKED ISSUE #3357: Implement differential rotation on the coordinate transform graph
During my recent work on reprojection, I was pondering how we could utilise the `reproject` package with it's different algorithms for resampling to our differential rotation warping code. One way to do this is to just swap out the call to `skimage.warp` with a call to `map_coordinates` inside reporoject (i.e using the [DeForest (2003)](https://www.boulder.swri.edu/~deforest/Papers_files/Sol%20Phys%202004%20Deforest.pdf)) algorithm.

What I am suggesting here is a potential way to be able to use `reproject` directly for this. To do this we would need to add the differential rotation models to the astropy transform graph. I think we could do this as one static `DifferentialRotationHeliographic` frame, which takes the rotation model name, new obstime, and new observer as parameters and then when transforming into this frame the differential rotation model would be applied. (I think that transforming out of it to HGS wouldn't do anything but I would need to think about that).

Having done this, all that would be needed to add to reproject is a `via=` keyword argument which provides an intermediate frame to transform through when doing the `pixel_to_pixel` transform.

I think implementing this like this would also make the logic of the differential rotation code a little simpler, as the observer corrections and other transform changes would become an implicit part of the transform graph, i.e if you wanted to transform from a HGS frame to a diff rot frame with a different observer the coordinate transform graph would do:
```
HGS(observer=obs_1, obstime=time_1) -> HGS(observer=obs_2, obstime=time_2) -> DiffRot(observer=obs_2, obstime=time_2)
```

@ayshih @wafels @wtbarnes

### PR DESCRIPTION
**Pull request by ayshih**
This PR integrates solar (differential) rotation into the coordinates framework.  The approach here is to enable defining a frame that is a modified version of a base coordinate frame by a duration of solar (differential) rotation.  The class factory `RotatedSunFrame` is supplied the "base" coordinate frame, and creates a base-specific rotated-Sun frame class.

Closes #3357

Still needs:
- [x] Add tutorial ([see this recent build](https://30287-2165383-gh.circle-artifacts.com/0/home/circleci/project/docs/_build/html/code_ref/coordinates/rotatedsunframe.html))
- [x] Finish writing tests
- [x] Support array `obstime`/`duration`/`rotated_time`

# Walkthrough
```python
>>> import astropy.units as u
>>> from sunpy.coordinates import RotatedSunFrame
>>> import sunpy.coordinates.frames as f
```

## Propagating a coordinate
Let's define a coordinate array of points in a rotated-Sun frame that in the base HGS frame were on the meridian as seen from Earth.  Here, the rotated-Sun frame applies +1 day of solar rotation.
```python
>>> hgs = f.HeliographicStonyhurst(obstime='2001-01-01')
>>> rotated_arc = RotatedSunFrame([0]*11*u.deg, range(-75, 90, 15)*u.deg, base=hgs, duration=1*u.day)
>>> rotated_arc
<RotatedSunHeliographicStonyhurst Coordinate (base=<HeliographicStonyhurst Frame (obstime=2001-01-01T00:00:00.000)>, duration=1.0 d, rotation_model=howard): (lon, lat, radius) in (deg, deg, km)
    [(0., -75., 695700.), (0., -60., 695700.), (0., -45., 695700.),
     (0., -30., 695700.), (0., -15., 695700.), (0.,   0., 695700.),
     (0.,  15., 695700.), (0.,  30., 695700.), (0.,  45., 695700.),
     (0.,  60., 695700.), (0.,  75., 695700.)]>
```
The coordinate components are relative to the distorted axes of the rotated-Sun frame, and thus are *already* differentially rotated in real space.  Convert the coordinate array back to the HGS frame to see what I mean.
```python
>>> rotated_arc.transform_to(hgs)
<HeliographicStonyhurst Coordinate (obstime=2001-01-01T00:00:00.000): (lon, lat, radius) in (deg, deg, km)
    [(10.7550473 , -75., 695700.), (11.70697161, -60., 695700.),
     (12.80904447, -45., 695700.), (13.68216339, -30., 695700.),
     (14.17617983, -15., 695700.), (14.32632838,   0., 695700.),
     (14.17617983,  15., 695700.), (13.68216339,  30., 695700.),
     (12.80904447,  45., 695700.), (11.70697161,  60., 695700.),
     (10.7550473 ,  75., 695700.)]>
```

##  Another example
There are several ways to instantiate a `RotatedSunFrame` frame.  Here, the input supplied to `base` has the relevant coordinate data, so that data is automatically "moved" to the correct place in the resulting frame.  Also, instead of supplying `duration` directly, the `rotated_time` keyword is used to indirectly define the duration of rotation.
```python
>>> center = f.Helioprojective(0*u.arcsec, 0*u.arcsec, obstime='2001-01-01', observer='earth')
>>> rotated_center = RotatedSunFrame(base=center, rotated_time='2001-01-03')
>>> rotated_center
<RotatedSunHelioprojective Coordinate (base=<Helioprojective Frame (obstime=2001-01-01T00:00:00.000, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate for 'earth'>)>, duration=2.0 d, rotation_model=howard): (Tx, Ty) in arcsec
    (0., 0.)>
```
No surprises here: The HPC center of the solar disk moves mostly in the X direction after +2 days of solar rotation.
```python
>>> rotated_center.transform_to(center)
<Helioprojective Coordinate (obstime=2001-01-01T00:00:00.000, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate for 'earth'>): (Tx, Ty, distance) in (arcsec, arcsec, AU)
    (468.86076705, -6.35238293, 0.97923501)>
```
Because `RotatedSunFrame` frames are in the transformation graph, you can do more creative transformations.  Here, we calculate what HPC coordinate when differentially rotated by -1 days corresponds to the disk center when differentially rotated by +2 days.
```python
>>> rotated_center.transform_to(RotatedSunFrame(base=center, duration=-1*u.day))
<RotatedSunHelioprojective Coordinate (base=<Helioprojective Frame (obstime=2001-01-01T00:00:00.000, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate for 'earth'>)>, duration=-1.0 d, rotation_model=howard): (Tx, Ty, distance) in (arcsec, arcsec, AU)
    (666.1751848, -13.91351551, 0.97991384)>
```

## Grid overlay on a Map
```python
>>> from sunpy.map import Map
>>> from sunpy.data.sample import AIA_171_IMAGE
>>> m = Map(AIA_171_IMAGE)

>>> import matplotlib.pyplot as plt
>>> ax = plt.subplot(projection=m)
>>> m.plot()

>>> overlay = ax.get_coords_overlay('heliographic_stonyhurst')
>>> overlay[0].set_ticks(spacing=15. * u.deg)
>>> overlay[1].set_ticks(spacing=90. * u.deg)
>>> overlay.grid(ls='-', color='white')

>>> overlay = ax.get_coords_overlay(RotatedSunFrame(base=f.HeliographicStonyhurst(obstime=m.date), duration=27*u.day))
>>> overlay[0].set_ticks(spacing=15. * u.deg)
>>> overlay[1].set_ticks(spacing=90. * u.deg)
>>> overlay.grid(ls='-', color='blue')
```
diffrot

### CODE REVIEW

**Comment by ayshih** [comment]
## Multiple rotation durations
You can work with array `obstime`/`duration`/`rotated_time`.
```python
>>> hpc = Helioprojective(obstime='2001-01-01', observer='earth')
>>> sc = SkyCoord([0]*9*u.arcsec, [0]*9*u.arcsec, frame=hpc)
>>> series = RotatedSunFrame(base=sc, duration=range(0, 27, 3)*u.day)
>>> series
<RotatedSunHelioprojective Coordinate (base=<Helioprojective Frame (obstime=2001-01-01T00:00:00.000, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate for 'earth'>)>, duration=[ 0.  3.  6.  9. 12. 15. 18. 21. 24.] d, rotation_model=howard): (Tx, Ty) in arcsec
    [(0., 0.), (0., 0.), (0., 0.), (0., 0.), (0., 0.), (0., 0.), (0., 0.),
     (0., 0.), (0., 0.)]>
>>> series.transform(hpc)
<Helioprojective Coordinate (obstime=2001-01-01T00:00:00.000, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate for 'earth'>): (Tx, Ty, distance) in (arcsec, arcsec, AU)
    [(   0.        ,    0.        , 0.97866506),
     ( 666.1751848 ,  -13.91351551, 0.97991384),
     ( 971.99406549,  -48.04083807, 0.98298359),
     ( 756.05760587,  -83.90802534, 0.98622017),
     ( 137.55563461, -102.40288732, 0.98789324),
     (-553.88198072,  -93.79075537, 0.98711382),
     (-951.06271076,  -62.59442403, 0.9842956 ),
     (-839.39922086,  -25.34560595, 0.98094109),
     (-274.84399172,   -2.09057135, 0.9788526 )]>
```

**Comment by ayshih** [comment]
## Transforming a coordinate to a frame with a different observation time
In the coordinates framework, a coordinate (e.g., `SkyCoord`) points to a specific location in inertial space, while the `obstime`/`observer` define the origin and axis orientations of the coordinate frame.  If you transform a `SkyCoord` to a coordinate frame with a different `obstime`, you obtain the *same* inertial location in the new coordinate frame, because there there isn't any way to specify how that coordinate evolves in time.

For example, here we try to transform disk center as seen by one observer to an observer 90 degrees separated in longitude and 6 days later in time:
```python
>>> observer1 = SkyCoord(HeliocentricInertial(0*u.deg, 0*u.deg, 1*u.AU, obstime='2001-01-01'))
>>> observer2 = SkyCoord(HeliocentricInertial(90*u.deg, 0*u.deg, 1*u.AU, obstime='2001-01-07'))

>>> frame1 = Helioprojective(observer=observer1, obstime=observer1.obstime)
>>> frame2 = Helioprojective(observer=observer2, obstime=observer2.obstime)

>>> sc = SkyCoord(0*u.arcsec, 0*u.arcsec, frame=frame1)
>>> sc.transform_to(frame2)
<SkyCoord (Helioprojective: obstime=2001-01-07T00:00:00.000, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate (obstime=2001-01-07T00:00:00.000): (lon, lat, radius) in (deg, deg, AU)
    (59.22396701, 0., 1.)>): (Tx, Ty, distance) in (arcsec, arcsec, AU)
    (-956.93749939, 1.16033026, 1.00006359)>
```
The output coordinate is on the east limb, which is consistent with not evolving the location with solar rotation.

With this PR, you can now include the evolution of solar rotation via this call:
```python
>>> RotatedSunFrame(base=sc, rotated_time=frame2.obstime).transform_to(frame2)
<Helioprojective Coordinate (obstime=2001-01-07T00:00:00.000, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate (obstime=2001-01-07T00:00:00.000): (lon, lat, radius) in (deg, deg, AU)
    (59.22396701, 0., 1.)>): (Tx, Ty, distance) in (arcsec, arcsec, AU)
    (-78.69026983, 1.16579504, 0.99537569)>
```
The `RotatedSunFrame` applies the differential rotation of the 6 days, and then the coordinate is transformed to the new coordinate frame.

File: sunpy/coordinates/metaframes.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    just that class, as well as ensuring that only one example of such a class
+    actually gets created in any given Python session.
+    """
+    # This code reuses significant code from Astropy's implementation of SkyOffsetFrame
+
+    if framecls in _rotatedsun_cache:
+        return _rotatedsun_cache[framecls]
+
+    # the class of a class object is the metaclass
+    framemeta = framecls.__class__
```
[the remark this answers is not in the document]
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: Sounds good, done

File: sunpy/coordinates/metaframes.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    base = Attribute()  # cannot use CoordinateAttribute here
+    duration = QuantityAttribute(default=0*u.day)
+    rotation_model = Attribute(default='howard')
+
+    def __new__(cls, *args, **kwargs):
+        # We don't want to call this method if we've already set up
+        # an rotated-Sun frame for this class.
+        if not (issubclass(cls, RotatedSunFrame) and cls is not RotatedSunFrame):
+            # We get the base argument, and handle it here.
+            try:
```
[the remark this answers is not in the document]
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: Sure.  This was copied from `SkyOffsetFrame` without much thought by me.

File: sunpy/coordinates/metaframes.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
+"""
+Coordinate frames that are defined relative to other frames
+"""
+
+from astropy import units as u
```
[the remark this answers is not in the document]
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: Done

File: sunpy/coordinates/metaframes.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+from .transformations import _transformation_debug
+from sunpy.time import parse_time
+from sunpy.time.time import _variables_for_parse_time_docstring
+from sunpy.util.decorators import add_common_docstring
+
+__all__ = ['NorthOffsetFrame', 'RotatedSunFrame']
+
+
+from .offset_frame import NorthOffsetFrame
+vars()['NorthOffsetFrame'].__module__ = __name__  # so that docs think that the class is local
```
[the remark this answers is not in the document]
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: Heh, yeah, I copied this style from something else I wrote, but it's unnecessary here.

**Comment by ayshih** [comment]
## Comparing reprojection using this frame versus the old approach

We can use this frame, plus `reproject`, to mimic the old approach to produce a differentially rotated map (`differential_rotate`).  This requires `reproject` 0.6 as well as PR #3540.  Importantly, the target frame for the reprojection needs to be the output frame differentially rotated to the time of the input frame.

comparison

```python
import matplotlib.pyplot as plt

from reproject import reproject_interp

from astropy.coordinates import SkyCoord
import astropy.units as u
import sunpy.map
from astropy.wcs import WCS
from sunpy.coordinates import RotatedSunFrame, get_body_heliographic_stonyhurst, frames as f
from sunpy.data.sample import AIA_171_IMAGE
from sunpy.physics.differential_rotation import differential_rotate

delta_t = 5*u.day

m = sunpy.map.Map(AIA_171_IMAGE)

# Define the output frame with an associated new observer (at Earth)
output_frame = f.Helioprojective(observer='earth', obstime=m.date + delta_t)

# Define the RotatedSunFrame (its `duration` will turn out to be -1 * `delta_t`)
rot_frame = RotatedSunFrame(base=output_frame, rotated_time=m.date)

out_shape = (1024, 1024)
header = sunpy.map.make_fitswcs_header(out_shape,
                                       SkyCoord(m.reference_coordinate.data, frame=output_frame),
                                       u.Quantity(m.reference_pixel),
                                       u.Quantity(m.scale))

out_wcs = WCS(header)
out_wcs.coordinate_frame = rot_frame

with sunpy.coordinates.transformations.ignore_sun_motion():
    arr, foot = reproject_interp(m, out_wcs, out_shape)

out_warp = sunpy.map.Map(arr, out_wcs)
out_warp.plot_settings = m.plot_settings

old = differential_rotate(m, time=delta_t)

fig = plt.figure(figsize=(15,8))
ax = plt.subplot(1,2,1, projection=out_warp)
out_warp.plot(ax)
ax.set_title("Reproject using RotatedSunFrame")
overlay = ax.get_coords_overlay(
    RotatedSunFrame(base=f.HeliographicStonyhurst(obstime=m.date), rotated_time=output_frame.obstime))
overlay[0].set_ticks(spacing=15. * u.deg)
overlay[1].set_ticks(spacing=90. * u.deg)
with sunpy.coordinates.transformations.ignore_sun_motion():
    overlay.grid(ls='-', color='blue')

ax2 = plt.subplot(1,2,2,projection=old)
old.plot(ax2)
ax2.set_title("Using differential_rotate")

plt.show()
```

**Comment by ayshih** [comment]
Now rebased off of #3540 so that I can use the context manager in examples.  Presumably #3540 will be merged before this PR.

**Comment by ayshih** [comment]
I've added a tutorial ([see this recent build](https://29503-2165383-gh.circle-artifacts.com/0/home/circleci/project/docs/_build/html/code_ref/coordinates/rotatedsunframe.html)).

**Comment by Cadair** [comment]
The tutorial is awesome! Are we confident in the reproject example doing the right thing?

**Comment by Cadair** [comment]
We should also open an issue on astropy about specifying the frame for a WCS. I spoke to both Tom and Erik about it at the astropy meeting and they both agreed it would be something useful to workout upstream.

**Comment by ayshih** [comment]
> The tutorial is awesome!

Thanks! I just added another example to show using `plot_coord` ([see recent build](https://29999-2165383-gh.circle-artifacts.com/0/home/circleci/project/docs/_build/html/code_ref/coordinates/rotatedsunframe.html#plotting-differentially-rotated-coordinates)).

> Are we confident in the reproject example doing the right thing?

Well, *I'm* confident that the coordinate transformations are correct.  Are *you* confident that `reproject` is doing the right thing? =P  Seriously though, the lingering discrepancies between the "new" method and the "old" method may be difficult to reconcile since we don't actually know that the "old" method was accurate.

File: docs/code_ref/coordinates/rotatedsunframe.rst [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+   from sunpy.map import Map
+
+   m = Map(AIA_171_IMAGE)
+
+   fig = plt.figure()
+   ax = plt.subplot(projection=m)
+   m.plot(clip_interval=(1., 99.95)*u.percent)
+
+   durations = range(-5, 6, 1)*u.day
+   point = SkyCoord([187]*len(durations)*u.arcsec, [283]*len(durations)*u.arcsec, frame=m.coordinate_frame)
```
Comment (Cadair) [inline docs/code_ref/coordinates/rotatedsunframe.rst]: The `*len(durations)` here is because of that astropy bug which stops the broadcasting working on obstime? 

I am wondering why we need to do that, and why specifying an array of `durations` for a singular coordinate point doesn't "just work".

Comment (ayshih) [reply docs/code_ref/coordinates/rotatedsunframe.rst]: Yeah, I can make this easier...

Comment (ayshih) [reply docs/code_ref/coordinates/rotatedsunframe.rst]: Okay, I've added automatic upgrading of the coordinate data, which makes the `plot_coord` example cleaner.

**Comment by Cadair** [comment]
@ayshih What's left on this before it's good to go?

**Comment by ayshih** [comment]
> @ayshih What's left on this before it's good to go?

I'm finalizing some hypothesis-based testing, although it's actually overkill here.  I'm not bothering to add explicit regression testing since honestly my tutorial will already do a thorough job of that through doctests.

**Comment by ayshih** [comment]
Okay, I've committed the test.  Who's brave enough to review this? =P

**Comment by Cadair** [comment]
Well it just took me 20 mins to read through your docs so...

File: sunpy/coordinates/metaframes.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    base frame. If such a class has already been created for this frame, the
+    same class will be returned.
+
+    This function is necessary because frame transformations depend
+    on connection between specific frame *classes*.  So each type of frame
+    needs its own distinct rotated-Sun frame class.  This function generates
+    just that class, as well as ensuring that only one example of such a class
+    actually gets created in any given Python session.
+    """
+    # This code reuses significant code from Astropy's implementation of SkyOffsetFrame
```
Comment (Cadair) [inline sunpy/coordinates/metaframes.py]:
```suggestion
    # This code reuses significant code from Astropy's implementation of SkyOffsetFrame
    # See licences/ASTROPY.rst
```

File: sunpy/coordinates/metaframes.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+    if framecls in _rotatedsun_cache:
+        return _rotatedsun_cache[framecls]
+
+    # the class of a class object is the metaclass
```
Comment (Cadair) [inline sunpy/coordinates/metaframes.py]:
```suggestion
    # Create a subclass of the frames metaclass, by getting the type of the class
    #  (which is the metaclass) and then subclassing it.
```

File: sunpy/coordinates/metaframes.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    framemeta = type(framecls)
+
+    class RotatedSunMeta(framemeta):
+        """
+        This metaclass renames the class to be "RotatedSun<framecls>".
+        """
+        def __new__(cls, name, bases, members):
+            # This has to be done because FrameMeta will set these attributes
```
Comment (Cadair) [inline sunpy/coordinates/metaframes.py]: If this comment is to believed why does this not need to be done after the call to `super()`?

**Comment by Cadair** [comment]
Now needs a rebase because #3540 has been merged.

File: sunpy/coordinates/metaframes.py (hunk 8 above)
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: I thought I understood this block (copied from `SkyOffsetFrame`), but now I'm not sure.  I'll do some testing.

File: sunpy/coordinates/metaframes.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+    Notes
+    -----
+    ``RotatedSunFrame`` is a factory class.  That is, the objects that it
+    yields are *not* actually objects of class ``RotatedSunFrame``.  Instead,
+    distinct classes are created on-the-fly for whatever the frame class is
+    of ``base``.
+    """
+    # This code reuses significant code from Astropy's implementation of SkyOffsetFrame
+
+    base = Attribute()  # cannot use CoordinateAttribute here
```
[the remark this answers is not in the document]
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: Yeah, I should

File: sunpy/coordinates/metaframes.py (hunk 8 above)
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: This code doesn't appear to be needed for the desired behavior (at least anymore), so may be vestigial in `SkyOffsetFrame`.  I'll remove it here in my next pass.

File: sunpy/coordinates/tests/test_metaframes.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    return {'hgs': f.HeliographicStonyhurst,
+            'hgc': f.HeliographicCarrington,
+            'hci': f.HeliocentricInertial,
+            'hcc': f.Heliocentric,
+            'hpc': f.Helioprojective,
+            }
+
+
+def test_class_creation(rot_frames, base_classes):
+    for rot_frame, base_class in [(rot_frames[key], base_classes[key]) for key in rot_frames.keys()]:
```
[the remark this answers is not in the document]
Comment (ayshih) [reply sunpy/coordinates/tests/test_metaframes.py]: Switched to an indirect-fixture implementation to get rid of this loop

File: sunpy/coordinates/metaframes.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: Note added

File: sunpy/coordinates/metaframes.py (hunk 7 above)
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: Comments added

File: sunpy/coordinates/metaframes.py (hunk 6 above)
Comment (ayshih) [reply sunpy/coordinates/metaframes.py]: Done

File: sunpy/coordinates/frames.py [hunk 11]
```diff
@@ -125,6 +125,14 @@ def represent_as(self, base, s='base', in_frame_units=False):
             data.lon.wrap_angle = self._wrap_angle
         return data
 
+    @property
+    def size(self):
+        """
+        Returns the size of the underlying data if it exists, else returns 0.  This overrides the
+        property in `~astropy.coordinates.BaseCoordinateFrame`.
+        """
+        return self.data.size if self.has_data else 0
```
Comment (dstansby) [inline sunpy/coordinates/frames.py]: Why this change?

File: docs/code_ref/coordinates/rotatedsunframe.rst [hunk 12]
```diff
@@ -0,0 +1,390 @@
+.. _sunpy-coordinates-rotatedsunframe:
+
+Differential rotation using coordinate frames
```
Comment (dstansby) [inline docs/code_ref/coordinates/rotatedsunframe.rst]: I think in this section there should be a clear reference to the formula (or paper that contains it) that is used for the differential rotaiton

File: docs/code_ref/coordinates/rotatedsunframe.rst [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+In a nutshell, the `~sunpy.coordinates.metaframes.RotatedSunFrame` class allows one to specify coordinates in a coordinate frame prior to an amount of solar (differential) rotation being applied.
+That is, the coordinate will point to a location in inertial space at some time, but will use a coordinate system at a *different* time to refer to that point, while accounting for the differential rotation between those two times.
+
+Technical note: ``RotatedSunFrame`` is not itself a coordinate frame, but is instead a "metaframe".
+A new frame class is created on the fly corresponding to each base coordinate frame class.
+This tutorial will refer to these new classes as ``RotatedSun*`` frames.
+
+Creating coordinates
+^^^^^^^^^^^^^^^^^^^^
+
```
Comment (dstansby) [inline docs/code_ref/coordinates/rotatedsunframe.rst]: There are lots of class links here that should be intersphinx links and not double-dashes. Happy for another issue to be opened to fix this though.

File: docs/code_ref/coordinates/rotatedsunframe.rst [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+There coordinates are stored in the ``RotatedSun*`` frame, but it can be useful to "pop off" this extra layer and retain only the coordinate representation in the base coordinate frame.
+There is a convenience method called `~sunpy.coordinates.metaframes.RotatedSunFrame.as_base()` to do exactly that.
+Be aware the resulting coordinate does *not* point to the same location in inertial space, despite the superficial similarity.
+Essentially, the component values have been copied from one coordinate frame to a different coordinate frame, and thus this is not merely a transformation between coordinate frames::
+
+  >>> new_coord.as_base()
+  <HeliographicCarrington Coordinate (obstime=2020-03-04T00:00:00.000): (lon, lat, radius) in (deg, deg, km)
+      (10., 20., 695700.)>
+
+Advanced uses of RotatedSunFrame
```
Comment (dstansby) [inline docs/code_ref/coordinates/rotatedsunframe.rst]: I think these examples would be better in individual gallery examples, since this is already quite an individually long bit of documentation.

File: sunpy/coordinates/frames.py (hunk 11 above)
Comment (ayshih) [reply sunpy/coordinates/frames.py]: For any frame with data, `astropy.coordinates` checks that any frame attributes that are non-scalar (i.e., size > 1) are able to be broadcast to the shape of the data.  See https://github.com/astropy/astropy/blob/03dfb7b5005f13bb975e6d6c5b0d60ca9e44a866/astropy/coordinates/baseframe.py#L614.

If a frame attribute is itself a frame, its size is checked to see if it's non-scalar.  However, the `BaseCoordinateFrame.size` property doesn't check whether the data exists before querying the data's size (see https://github.com/astropy/astropy/blob/03dfb7b5005f13bb975e6d6c5b0d60ca9e44a866/astropy/coordinates/baseframe.py#L707).  That means that a data-less frame raises an error when used as an attribute for a frame with data.

In this implementation/API, I specifically need to be able to precisely that.  That's why I modified `SunPyBaseCoordinateFrame.size` to return zero if there is no data.  As long as code doesn't query a frame's size until after it confirms that there is frame data, there's no difference in behavior.  It's such a subtle issue that I've haven't yet bothered opening an issue upstream to discuss it.

**Comment by Cadair** [comment]
@dstansby or @ayshih can you open an issue to track the documentation improvements please?

**Comment by dstansby** [comment]
👍 will do
